File: rust/src/valid_parentheses.rs
use std::collections::HashMap;

pub fn is_valid(s: String) -> bool {
    let mut o_c_map = HashMap::with_capacity(3);
    o_c_map.insert(')', '(');
    o_c_map.insert('}', '{');
    o_c_map.insert(']', '[');
    let mut stack = vec![];

    for c in s.chars() {
        if o_c_map.contains_key(&c) {
            match stack.pop() {
                Some(e) => {
                    if o_c_map[&c] != e {
                        return false;
                    }
                }
                None => return false,
            }
        } else {
            stack.push(c);
        }
    }

    stack.is_empty()
}

File: python/valid_parentheses.py
def isValid(s: str) -> bool:
    grouping_symbols = {
        '}': '{',
        ')': '(',
        ']': '['
    }

    stack = []
    for c in s:
        if c not in grouping_symbols:
            stack.append(c)
        else:
            top = stack.pop() if stack else '@'
            if grouping_symbols[c] != top:
                return False

    return not stack
